helix-view/src/document.rs

```diff
@@ -548,6 +561,9 @@ impl Document {
         // TODO: mark changes up to now as saved
 
         let language_server = self.language_server.clone();
+        if let Some(vcs) = &mut self.version_control {
+            vcs.diff();
+        }
```

Comment:
I think the bug that the diff showed out too late is probably because `save_impl` is asynchronous and this does not triggers render after it is finished.

Reply:
Any suggestions on how to fix this?

Reply:
Ok, I see now that the diff occurs before the async save call, which means it occurs before the file is actually written.
I'm not sure how to fix it in an elegant way yet, but there is progress.

Reply:
Sorry, forgot to comment here, I think the way to go would be to `.and_then` the save future, in that closure schedule another callback (using something similar as [cx.callback](https://github.com/helix-editor/helix/blob/48e344a2a813b875ca813be726387692b7ba0762/helix-term/src/commands.rs#L92-L111)) and modify the document in there.

I'm planning on making this a bit easier to do by introducing hooks, so saving would trigger a `hook::DOCUMENT_SAVED` and schedule jobs like these.

Reply:
I had some time to look at it again.

From what I can see, we need some way to have a job queue with access to the document and which the document can access.

I'm thinking of giving document a jobs queue. This would be similar to how Compositor polls it's layers, now Editor will poll the documents for any jobs.

This would allow us to move several other functions to async, like possibly open and format.

Reply:
I think we could maybe make the existing jobs queue accessible globally, and in the callback you can simply use a passed in id to find the document (`editor.documents[id]`)

Reply:
Format already uses the callback system after some of the recent PRs, it receives the changes in a job then applies them to the doc: https://github.com/helix-editor/helix/pull/285